### ISSUE TITLE
Q: OpenGL support

### ISSUE THREAD

**Issue by Eric89GXL** [pybee/toga#51]
Is there any plan to support OpenGL contexts / widgets in `toga`? There is code for creating them for OSX/Linux/Windows in `Pyglet`, so there is some guideline available. Just curious to know if I should keep an eye on it, because if OpenGL contexts are available then it could be made to operate with / as a backend for [vispy] (http://github.com/vispy/vispy).

**Comment by freakboy3742** [comment]
Eventually - sure. It's not a high priority for myself personally, but I know 3D visualizations can be very significant for others, and most platforms are now providing relatively simple canvases for OpenGL. So, if some widgets were to mysteriously appear in the pull request queue... :-)

Completely seriously, it's actually not as hard to write a Toga widget as you may think - if there's a good base widget with an obvious API to expose, writing the wrapper API is a relatively simple proposition. The code for the Button widget is a good place to start if you're interested in contributing.

**Comment by CharsargCodeDevelopment** [comment]
Any progress?

**Comment by mhsmith** [comment]
No, but if someone contributed this feature, we'd be happy to accept it.

**Comment by corranwebster** [comment]
I have a very rough proof of concept of an OpenGL widget in #4263.

A design discussion is needed.

### Supporting Backends

Currently Qt and Android are supported in the prototype, because those are comparatively straightforward in exposing an OpenGL widget *and* an OpenGL API as part of the native library. ~I *think* iOS might be similar.~ (iOS has OpenGL ES exposed as-is, so it is more like Cocoa and Gtk.) And for the Web, WebGL is fairly straightforward to get access to.

Cocoa and Gtk don't include OpenGL bindings as part of their APIs, expecting programs in ObjC or C++ to call OpenGL directly via the C API once the context is set correctly. Here there is a question about whether to wrap the bits of OpenGL we need ad-hoc using `ctypes`, or whether to use something like PyOpenGL (or perhaps repurpose bits of Pyglet). If we wrap using `ctypes` then I suspect a lot of the code will be common between the two backends.

For Winforms I'm not sure of the best approach: OpenTK has C# bindings that work with Winforms; on the other hand working directly with the C API may be simpler.

Note also that macOS and iOS have officially deprecated OpenGL, and want people to use Metal. There are projects like ANGLE that provide adaptive layers, and it is possible this may be a better long-term approach. There is a similar question around Vulkan on other platforms.

### Interface Design

I think the goal of the Toga interface should be to support a modern subset of OpenGL using shader-based rendering. I *think* (although I could be convinced otherwise by people more familiar with OpenGL) that just like we are using a "pythonified" version of the HTML Canvas as a guide for the API of the `Canvas` widget, aiming for a "pythonified" version of the relevant bits of the WebGL API may be worthwhile. In particular, I think it should look something like:

- the user supplies a redraw callback (or possibly several, may need an `init` callback too, for example) to the `OpenGLView` interface which is called with an "OpenGLContext" instance which exposes the various OpenGL methods that we want to support, but with PEP 8-style names (so `glClearColor` becomes `clear_color`, etc.)
- this can, I think, be an object supplied by the implementation that satisfies a protocol: I don't think we need a formal interface/implementation split because the underlying OpenGL API is very similar across backends. (This would be like having the Canvas supply the implementation's new `Context` classes directly to the drawing library.
- it is likely that the interface will want convenience APIs for things like shaders, programs, textures, buffers, etc. as well
- it is possible that there should be a high-level interface like Canvas has, but I don't know what that would look like: to be useful it would need good design.

One other possibility is to use something like [ModernGL](https://moderngl.readthedocs.io/en/5.8.2/) to provide a more Pythonic API, but I'm not sure how well that will integrate with the Qt, iOS or Android backends.

**Comment by corranwebster** [comment]
More exploration:
- the PySide6 bindings for the OpenGL backend have some issues (eg. `glShaderiv()` doesn't seem bound correctly to return a value).
- I think a big question is about how to provide the OpenGL ctypes bindings - these are going to be needed for cocoa, iOS and Gtk, and may be needed to patch some Qt methods - there are XML descriptions of the APIs and Pyglet has tools for autogenerating ctypes bindings from them. Plus there's always PyOpenGL (not sure if it supports iOS, though).
- there is a question of what OpenGL versions to support; this is important for the shader languages more than the C API, I think.

**Comment by corranwebster** [comment]
> * I think a big question is about how to provide the OpenGL ctypes bindings - these are going to be needed for cocoa, iOS and Gtk, and may be needed to patch some Qt methods - there are XML descriptions of the APIs and Pyglet has tools for autogenerating ctypes bindings from them. Plus there's always PyOpenGL (not sure if it supports iOS, though).

Some investigation here:
- PyOpenGL is not supported on iOS. I *think* it could be, but it would require a new platform to be added, plus PyOpenGL_accelerate doesn't have iOS wheels, and there may be work needed to support that.
- PyOpenGL source code hints that it may work on Android (after all, Android is a linux variant), but quick experiments didn't work out.
- PyOpenGL seems to work on Cocoa and Qt (at least on Mac) and I would expect it to work with Gtk and probably windows.

So it would seem that:
- iOS support needs an OpenGLES wrapper, probably using ctypes. Pyglet's auto-generated wrappers (or the wrapper generator) seem like the best options, but we *could* manually wrap the minimum amount needed.
- Android is probably easiest to go through `android.opengl` even though that's through a Java API
- For Cocoa, if we're automatically wrapping for iOS, the same tooling would likely work for Cocoa
- For fast development, it probably makes sense to implement on top of PyOpenGL first, and then worry about the other backends and whether we want to continue with PyOpenGL long-term.

**Comment by corranwebster** [comment]
More thought on the big picture for building this. Perhaps a two-phase approach might work:

1. Create widgets which call back to user code with the OpenGL view configured and ready to draw, but leave it up to the user to manage the actual OpenGL drawing completely. Basically it's BYO OpenGL, and the API is up to the user - PyOpenGL, ModernGL, android.opengl, Qt's OpenGL library, WebGL, custom wrappers they wrote themselves, whatever.  The OpenGL code won't be super-portable, particularly for web and mobile, but it gets something out the door and working.
 
2.  A Pythonic, WebGL-ish cross platform API that permits something close to common code to be written for all backends. ModernGL's API may be an inspiration for what this looks like as well. This API only supports a modern subset of OpenGL for shader-based rendering, so it is optional in the sense that users can still choose to use their own preferred OpenGL API as long as it works on their platform. This also means the API can be built incrementally, adding basic functionality but where there are gaps the low-level APIs can be used to fill them at the cost of portability.

I think this approach has a few of advantages

- it makes integration with existing OpenGL code easier, eg. VisPy as mentioned early in this issue, which has it's own OpenGL API layer, Gloo.
- the more user-friendly API makes things more approachable for users less familiar with OpenGL.
- expert users can still access full OpenGL if they need it.
- implementing it is more tractable than a monolithic "get it right first time" API (which honestly is daunting!)
